daemon/cmd/daemon_main.go

```diff
@@ -1768,31 +1700,6 @@ func runDaemon() {
 	}
 	bootstrapStats.k8sInit.End(true)
 	restoreComplete := d.initRestore(restoredEndpoints)
-	if wgAgent != nil {
-		if err := wgAgent.RestoreFinished(); err != nil {
```

Comment:
The more I look at this existing code, the more I wonder if multiple of these steps are assuming that restore is complete rather than checking the `<-d.initRestore(..)`. cc @gandro

Reply:
Not sure I fully understand the question, but the reason `Init` and `RestoreFinished` are separate is that we call `Init` to be able to get a reference we can hand out to the IPCache and NodeManager. Once all remote endpoints have been discovered, we call `RestoreFinished`, at which point we re-initialize the WireGuard device with all remote endpoints that exist after the restore. 

This is not well documented, all that I found is this comment here:
https://github.com/cilium/cilium/blob/130506e63f66bcb73339aecb857dbd374ef9094b/pkg/wireguard/agent/agent.go#L57-L61

So what WireGuard needs is the public key and endpoints of all remote nodes to be available. So what we are actually waiting on here is `<-d.k8sCachesSynced` (a few lines above). I'm actually not sure if `<-d.k8sCachesSynced` guarantees us that the node manager has processed all nodes in kvstore, so that might be a bug in the old code. Either way, the new version is also not correct either, as I don't see any code waiting for the "all remote nodes have been discovered" event.

(side-note: For IPAM, `RestoreFinished` has a completely different meaning. There we actually need to wait for local endpoint restoration, which is a different event. We should likely rename the WireGuard method to reflect that)

Reply:
I think we'll want to make these very explicit when we refactor the code, e.g. actually make wireguard depend on this condition explicitly, and I think ideally it wouldn't depend on k8s directly, but rather on the node manager to become ready, either by making the startup of node manager block, or making all calls to node manager block.